saliency/xrai.py

```diff
@@ -0,0 +1,555 @@
+                  information about multiple segments by encoding them
+                  with distinct integer values. If the value is None,
+                  a defaut segmentation algorithm will be applied. Defaults to
+                  None.
+        extra_parameters: an XRAIParameters object that specifies
+                          additional parameters for the XRAI saliency
+                          method. Defaults to None.
+
+    Raises:
+        ValueError: If algorithm type is unknown (not full or fast)
+
+    Returns:
+        XRAIOutput: an object that contains the output of the XRAI algorithm.
+
+    TODO(tolgab) Add output_selector functionality from XRAI API doc
+    """
+    if extra_parameters.verbosity > 1:
+      logging.info("Computing IG...")
+    x_baselines = self._make_baselines(x_value, baselines)
+
+    attrs = self._get_integrated_gradients(x_value,
+                                           feed_dict=feed_dict,
+                                           baselines=x_baselines,
+                                           steps=extra_parameters.steps)
+    # Merge attributions from different baselines
+    attr = np.mean(attrs, axis=0)
+    # Merge attribution channels for XRAI input
+    attr = _attr_aggregation_max(attr)
+
+    if extra_parameters.verbosity > 1:
+      logging.info("Done with IG. Computing XRAI...")
+    if segments is not None:
+      segs = segments
+    else:
+      segs = _get_segments_felsenschwab(x_value)
+
+    if extra_parameters.algorithm == 'full':
+      attr_map, attr_data = self._xrai(
+          attr=attr,
+          segs=segs,
+          area_perc_th=extra_parameters.area_threshold,
+          gain_fun=_gain_density,
+          verbose=extra_parameters.verbosity,
+          integer_segments=extra_parameters.flatten_xrai_segments)
+    elif extra_parameters.algorithm == 'fast':
+      attr_map, attr_data = self._xrai_fast(
+          attr=attr,
+          segs=segs,
+          gain_fun=_gain_density,
+          verbose=extra_parameters.verbosity,
+          integer_segments=extra_parameters.flatten_xrai_segments)
+    else:
+      logging.error('Unknown algorithm type: {}'.format(
+          extra_parameters.algorithm))
+      raise ValueError
+
+    results = XRAIOutput(attr_map)
+    results.baselines = x_baselines
+    if extra_parameters.return_xrai_segments:
+      results.segments = attr_data
+    if extra_parameters.return_baseline_predictions:
+      baseline_predictions = []
+      for baseline in x_baselines:
+        baseline_predictions.append(self._predict(baseline))
+      results.baseline_predictions = baseline_predictions
+    if extra_parameters.return_ig_attributions:
+      results.ig_attribution = attr
+    return results
+
+  @staticmethod
+  def _xrai(attr,
+            segs,
+            gain_fun=_gain_density,
+            area_perc_th=1.0,
+            verbose=0,
+            min_pixel_diff=50,
```

Comment:
added min_pixel_diff as an experimental parameter to XRAIParameters for now so people can edit it in case they have smaller segments. Deferring anything related to min segment size to future PR.